src/function.rs

```diff
@@ -246,28 +250,44 @@ where
 
     /// True if the input `input` contains a memo that cites itself as a cycle head.
     /// This indicates an intermediate value for a cycle that has not yet reached a fixed point.
-    fn cycle_head_kind(&self, zalsa: &Zalsa, input: Id) -> CycleHeadKind {
+    fn cycle_head_kind(
+        &self,
+        zalsa: &Zalsa,
+        input: Id,
+        iteration: Option<IterationCount>,
+    ) -> CycleHeadKind {
         let is_provisional = self
             .get_memo_from_table_for(zalsa, input, self.memo_ingredient_index(zalsa, input))
             .is_some_and(|memo| {
-                memo.cycle_heads()
-                    .into_iter()
-                    .any(|head| head.database_key_index == self.database_key_index(input))
+                Some(memo.revisions.iteration()) != iteration
+                    || !memo.revisions.verified_final.load(Ordering::Relaxed)
```

Comment:
If the memo is verified final, but the iteration doesn't match, we still want to report it as provisional? This seems surprising to me, that we would ever want to report a verified-final memo as provisional.

Reply:
I think what we should do here is to change the return type to include the iteration for a final memo. I believe I added a comment at the call site why it's important (we may have a memo that is provisional and depends on a non-final iteration memo because the final iteration didn't depend on that memo. We should not finalize this memo)